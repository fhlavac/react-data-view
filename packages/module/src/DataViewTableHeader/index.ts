export { default } from './DataViewTableHeader';
export * from './DataViewTableHeader';
