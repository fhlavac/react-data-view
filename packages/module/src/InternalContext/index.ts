export { default } from './InternalContext';
export * from './InternalContext';
