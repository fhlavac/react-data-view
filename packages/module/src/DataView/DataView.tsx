import React from 'react';
import { Stack, StackItem } from '@patternfly/react-core';
import { DataViewSelection, InternalContextProvider } from '../InternalContext';

export const DataViewState = {
  empty: 'empty',
  loading: 'loading',
  error: 'error'
} as const;

export type DataViewState = typeof DataViewState[keyof typeof DataViewState];

export interface DataViewProps {
  /** Content rendered inside the data view */
  children: React.ReactNode;
  /** Custom OUIA ID */
  ouiaId?: string;
  /** Selection context configuration */
  selection?: DataViewSelection;
  /** Currently active state */
  activeState?: DataViewState | string;
}

export type DataViewImpementationProps = Omit<DataViewProps, 'onSelect' | 'isItemSelected' | 'isItemSelectDisabled'>;

const DataViewImplementation: React.FC<DataViewImpementationProps> = ({
  children, ouiaId = 'DataView', ...props
}: DataViewImpementationProps) => (
  <Stack data-ouia-component-id={`${ouiaId}-stack`} {...props}>
    {React.Children.map(children, (child, index) => (
      <StackItem data-ouia-component-id={`${ouiaId}-stack-item-${index}`}>
        {child}
      </StackItem>
    ))} 
  </Stack>
)

export const DataView: React.FC<DataViewProps> = ({ children, selection, activeState, ...props }: DataViewProps) => (
  <InternalContextProvider selection={selection} activeState={activeState} >
    <DataViewImplementation {...props}>{children}</DataViewImplementation>
  </InternalContextProvider>
);

export default DataView;
