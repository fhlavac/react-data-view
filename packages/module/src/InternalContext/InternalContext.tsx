import React, { createContext, PropsWithChildren, useContext } from 'react';
import { DataViewState } from '../DataView';

export interface DataViewSelection {
  /** Called when the selection of items changes */
  onSelect: (isSelecting: boolean, items?: any[] | any) => void;  // eslint-disable-line @typescript-eslint/no-explicit-any
  /** Checks if a specific item is currently selected */
  isSelected: (item: any) => boolean;  // eslint-disable-line @typescript-eslint/no-explicit-any
  /** Determines if selection is disabled for a given item */
  isSelectDisabled?: (item: any) => boolean;  // eslint-disable-line @typescript-eslint/no-explicit-any
}

export interface InternalContextValue {
  selection?: DataViewSelection;
  activeState?: DataViewState; 
}

export const InternalContext = createContext<InternalContextValue>({
  selection: undefined,
  activeState: undefined
});

export type InternalProviderProps = PropsWithChildren<InternalContextValue>

export const InternalContextProvider: React.FC<InternalProviderProps> = ({
  children,
  selection,
  activeState
}) => (
  <InternalContext.Provider
    value={{ selection, activeState }}
  >
    {children}
  </InternalContext.Provider>
);

export const useInternalContext = () => useContext(InternalContext);

export default InternalContext;
