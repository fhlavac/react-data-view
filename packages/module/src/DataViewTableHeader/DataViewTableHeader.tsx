import React from 'react';
import {
  Th,
  Thead,
  TheadProps,
  Tr
} from '@patternfly/react-table';
import { useInternalContext } from '../InternalContext';
import { DataViewTh, isDataViewThObject } from '../DataViewTable';

export interface DataViewTableHeaderProps extends TheadProps {
  /** Indicates whether table is a tree */
  isTreeTable?: boolean;
  /** Columns definition */
  columns: DataViewTh[];
  /** Custom OUIA ID */
  ouiaId?: string;
}

export const DataViewTableHeader: React.FC<DataViewTableHeaderProps> = ({
  isTreeTable = false,
  columns,
  ouiaId = 'DataViewTableHeader',
  ...props
}: DataViewTableHeaderProps) => {
  const { selection } = useInternalContext();
  const { onSelect, isSelected } = selection ?? {};

  return (
    <Thead data-ouia-component-id={`${ouiaId}-thead`} {...props}>
      <Tr ouiaId={`${ouiaId}-tr-head`}>
        {onSelect && isSelected && !isTreeTable && <Th key="row-select" screenReaderText='Data selection table header cell' />}
        {columns.map((column, index) => (
          <Th
            key={index}
            {...(isDataViewThObject(column) && (column?.props ?? {}))}
            data-ouia-component-id={`${ouiaId}-th-${index}`}
          >
            {isDataViewThObject(column) ? column.cell : column}
          </Th>
        ))}
      </Tr>
    </Thead>
  );
};

export default DataViewTableHeader;
