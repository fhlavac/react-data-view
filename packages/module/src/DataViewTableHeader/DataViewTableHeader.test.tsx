import React from 'react';
import { render } from '@testing-library/react';
import { Table } from '@patternfly/react-table';
import { DataViewTableHeader } from './DataViewTableHeader';
import { DataViewSelection } from '../InternalContext';
import { DataView } from '../DataView';

const columns = [ 'Repositories', 'Branches', 'Pull requests', 'Workspaces', 'Last commit' ];

const ouiaId = 'HeaderExample';

describe('DataViewTableHeader component', () => {
  const mockSelection: DataViewSelection = {
    onSelect: jest.fn(),
    isSelected: jest.fn(),
    isSelectDisabled: jest.fn(),
  };

  test('should render correctly', () => {
    const { container } = render(
      <Table>
        <DataViewTableHeader columns={columns} ouiaId={ouiaId} />
      </Table>
    );
    expect(container).toMatchSnapshot();
  });

  test('should render the tree table correctly when isTreeTable is true', () => {
    const { container } = render(
      <DataView selection={mockSelection}>
        <Table>
          <DataViewTableHeader isTreeTable columns={columns} ouiaId={ouiaId} />
        </Table>
      </DataView>
    );
    expect(container).toMatchSnapshot();
  });

  test('should render selection column when selection is provided', () => {
    const { container } = render(
      <DataView selection={mockSelection}>
        <Table>
          <DataViewTableHeader columns={columns} ouiaId={ouiaId} />
        </Table>
      </DataView>
    );
    expect(container).toMatchSnapshot();
  });
});



