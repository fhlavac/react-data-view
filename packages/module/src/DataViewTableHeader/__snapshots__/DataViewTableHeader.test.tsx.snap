// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataViewTableHeader component should render correctly 1`] = `
<div>
  <table
    class="pf-v5-c-table pf-m-grid-md"
    data-ouia-component-id="OUIA-Generated-Table-1"
    data-ouia-component-type="PF5/Table"
    data-ouia-safe="true"
    role="grid"
  >
    <thead
      class="pf-v5-c-table__thead"
      data-ouia-component-id="HeaderExample-thead"
    >
      <tr
        class="pf-v5-c-table__tr"
        data-ouia-component-id="HeaderExample-tr-head"
        data-ouia-component-type="PF5/TableRow"
        data-ouia-safe="true"
      >
        <th
          class="pf-v5-c-table__th"
          data-ouia-component-id="HeaderExample-th-0"
          scope="col"
          tabindex="-1"
        >
          Repositories
        </th>
        <th
          class="pf-v5-c-table__th"
          data-ouia-component-id="HeaderExample-th-1"
          scope="col"
          tabindex="-1"
        >
          Branches
        </th>
        <th
          class="pf-v5-c-table__th"
          data-ouia-component-id="HeaderExample-th-2"
          scope="col"
          tabindex="-1"
        >
          Pull requests
        </th>
        <th
          class="pf-v5-c-table__th"
          data-ouia-component-id="HeaderExample-th-3"
          scope="col"
          tabindex="-1"
        >
          Workspaces
        </th>
        <th
          class="pf-v5-c-table__th"
          data-ouia-component-id="HeaderExample-th-4"
          scope="col"
          tabindex="-1"
        >
          Last commit
        </th>
      </tr>
    </thead>
  </table>
</div>
`;

exports[`DataViewTableHeader component should render selection column when selection is provided 1`] = `
<div>
  <div
    class="pf-v5-l-stack"
    data-ouia-component-id="DataView-stack}"
  >
    <div
      class="pf-v5-l-stack__item"
      data-ouia-component-id="DataView-stack-item-0"
    >
      <table
        class="pf-v5-c-table pf-m-grid-md"
        data-ouia-component-id="OUIA-Generated-Table-3"
        data-ouia-component-type="PF5/Table"
        data-ouia-safe="true"
        role="grid"
      >
        <thead
          class="pf-v5-c-table__thead"
          data-ouia-component-id="HeaderExample-thead"
        >
          <tr
            class="pf-v5-c-table__tr"
            data-ouia-component-id="HeaderExample-tr-head"
            data-ouia-component-type="PF5/TableRow"
            data-ouia-safe="true"
          >
            <th
              class="pf-v5-c-table__th"
              scope="col"
              tabindex="-1"
            >
              <span
                class="pf-v5-screen-reader"
              >
                Data selection table header cell
              </span>
            </th>
            <th
              class="pf-v5-c-table__th"
              data-ouia-component-id="HeaderExample-th-0"
              scope="col"
              tabindex="-1"
            >
              Repositories
            </th>
            <th
              class="pf-v5-c-table__th"
              data-ouia-component-id="HeaderExample-th-1"
              scope="col"
              tabindex="-1"
            >
              Branches
            </th>
            <th
              class="pf-v5-c-table__th"
              data-ouia-component-id="HeaderExample-th-2"
              scope="col"
              tabindex="-1"
            >
              Pull requests
            </th>
            <th
              class="pf-v5-c-table__th"
              data-ouia-component-id="HeaderExample-th-3"
              scope="col"
              tabindex="-1"
            >
              Workspaces
            </th>
            <th
              class="pf-v5-c-table__th"
              data-ouia-component-id="HeaderExample-th-4"
              scope="col"
              tabindex="-1"
            >
              Last commit
            </th>
          </tr>
        </thead>
      </table>
    </div>
  </div>
</div>
`;

exports[`DataViewTableHeader component should render the tree table correctly when isTreeTable is true 1`] = `
<div>
  <div
    class="pf-v5-l-stack"
    data-ouia-component-id="DataView-stack}"
  >
    <div
      class="pf-v5-l-stack__item"
      data-ouia-component-id="DataView-stack-item-0"
    >
      <table
        class="pf-v5-c-table pf-m-grid-md"
        data-ouia-component-id="OUIA-Generated-Table-2"
        data-ouia-component-type="PF5/Table"
        data-ouia-safe="true"
        role="grid"
      >
        <thead
          class="pf-v5-c-table__thead"
          data-ouia-component-id="HeaderExample-thead"
        >
          <tr
            class="pf-v5-c-table__tr"
            data-ouia-component-id="HeaderExample-tr-head"
            data-ouia-component-type="PF5/TableRow"
            data-ouia-safe="true"
          >
            <th
              class="pf-v5-c-table__th"
              data-ouia-component-id="HeaderExample-th-0"
              scope="col"
              tabindex="-1"
            >
              Repositories
            </th>
            <th
              class="pf-v5-c-table__th"
              data-ouia-component-id="HeaderExample-th-1"
              scope="col"
              tabindex="-1"
            >
              Branches
            </th>
            <th
              class="pf-v5-c-table__th"
              data-ouia-component-id="HeaderExample-th-2"
              scope="col"
              tabindex="-1"
            >
              Pull requests
            </th>
            <th
              class="pf-v5-c-table__th"
              data-ouia-component-id="HeaderExample-th-3"
              scope="col"
              tabindex="-1"
            >
              Workspaces
            </th>
            <th
              class="pf-v5-c-table__th"
              data-ouia-component-id="HeaderExample-th-4"
              scope="col"
              tabindex="-1"
            >
              Last commit
            </th>
          </tr>
        </thead>
      </table>
    </div>
  </div>
</div>
`;
